test/e2e/serial/tests/tolerations.go

```diff
@@ -783,10 +827,51 @@ func sriovToleration() corev1.Toleration {
 	}
 }
 
-func waitForMcpUpdate(cli client.Client, ctx context.Context, mcps []*machineconfigv1.MachineConfigPool) {
-	By("waiting for mcp to start updating")
-	Expect(waitForMcpsCondition(cli, ctx, mcps, machineconfigv1.MachineConfigPoolUpdating)).To(Succeed())
+func waitForMcpUpdate(cli client.Client, ctx context.Context, mcpsInfo []mcpInfo, updateType MCPUpdateType) {
+	for _, info := range mcpsInfo {
+		By(fmt.Sprintf("verify updates for mcp %q", info.obj.Name))
+		klog.Info("waiting for mcp to start updating")
+		if err := waitForMcpsCondition(cli, ctx, []*machineconfigv1.MachineConfigPool{info.obj}, machineconfigv1.MachineConfigPoolUpdating); err != nil {
+			klog.Warningf("failed to find mcps while in updating status")
+		}
+
+		klog.Info("wait for mcp to get updated")
+		//here we must fail on errors
+		Expect(waitForMcpsCondition(cli, ctx, []*machineconfigv1.MachineConfigPool{info.obj}, machineconfigv1.MachineConfigPoolUpdated)).To(Succeed())
+
+		var updatedMcp machineconfigv1.MachineConfigPool
+		Expect(cli.Get(ctx, client.ObjectKeyFromObject(info.obj), &updatedMcp)).To(Succeed())
+		// Note: don't check for difference between initial config and current config on the updated mcp because mcp going into
+		// an update doesn't always it goes into a configuration update and thus associated to different MC, it could be because
+		// new nodes are joining the pool so the MC update is happening on those nodes
+		updatedConfig := updatedMcp.Status.Configuration.Name
+		if updateType == MachineConfig {
+			Expect(updatedConfig).ToNot(Equal(info.initialConfig))
+		}
+
+		// MachineConfig update type will also update the node currentConfig so check that anyway
+		klog.Info("verify mcp config is updated by ensuring the sample node has updated MC")
+		ok, err := verifyUpdatedMCOnNodes(cli, ctx, info.sampleNode, updatedConfig)
+		Expect(err).ToNot(HaveOccurred())
+		Expect(ok).To(BeTrue())
+	}
```

Comment:
This should be done in a loop, which should periodically poll the given MCP and exit succesfully if all the following are true:
- condition is Updated
- the desired config matches the expected value or, equivalently but less strong (but way easier to check) is different from the initial state

Reply:
the loop is done on the Updated condition `waitForMcpsCondition` (and before that another loop is done on Updating). the desired config on the node being updated to the new state is a result of fulfilling the Updated condition. so if the mcp Updated==true only, then it is reasonable to verify the node's current/desired status. Even if the specific node status is updated, we still have work to do, so the deciding factor here is the mcp Updated condition. We can put it in the loop alongside the Updated condition, but if the node status passed, it will not make any difference if, on the other hand, the Updated is still false, so what we mainly check for here is Updated=true, and only then the other conditions would be legit to check. 
If the concern is that the node's current vs desired config would take time to be updated after the mcp is done updating, we can make verifyUpdatedMCOnNodes functionality in eventually but as far as I learned mcp will not turn Updated unless all nodes got the correct annotations, otherwise the node should be reported as degraded and the reason would not be empty, which will make the mcp status Updated=false